src/main/java/org/apache/sysml/hops/ipa/InterProceduralAnalysis.java

```diff
@@ -543,7 +546,27 @@ else if (sb instanceof ForStatementBlock) //incl parfor
 			propagateStatisticsIntoFunctions(prog, roots, fcand, callVars, fcandSafeNNZ, unaryFcands, fnStack);
 		}
 	}
-	
+
+	/**
+	 * Propagate scalar values across DAGs.
+	 *
+	 * This replaces scalar transient reads with literals.
+	 *
+	 * @param roots  List of HOPs.
+	 * @param vars  Map of variables eligible for propagation.
+	 * @throws HopsException
+	 */
+	private void propagateScalarsAcrossDAG(ArrayList<Hop> roots, LocalVariableMap vars)
+		throws HopsException
+	{
+		for (Hop hop : roots) {
+			try {
+				Recompiler.rReplaceLiterals(hop, vars);
+			} catch (Exception ex) {
+				throw new HopsException("Failed to replace transient reads w/ literals.", ex);
```

Comment:
Maybe use a slightly more general exception message as literal replacement affects not only transient reads, but also value types casts on top of them.